iree/compiler/Dialect/HAL/Transforms/PublicAbiGeneration.cpp

```diff
@@ -216,6 +216,44 @@ LogicalResult generateSynchronousBody(
   return success();
 }
 
+LogicalResult generateAsynchronousBody(FuncOp funcOp, FuncOp syncFuncOp,
+                                       OpBuilder moduleBuilder) {
+  auto *ctx = funcOp.getContext();
+  auto loc = funcOp.getLoc();
+  Block *entryBlock = funcOp.addEntryBlock();
+  OpBuilder builder = OpBuilder::atBlockEnd(entryBlock);
+
+  // Wait until the wait semaphore reaches the wait value.
+  auto waitSemaphore = entryBlock->getArgument(0);
+  auto waitValue = entryBlock->getArgument(1);
+  auto waitOp = builder.create<HAL::SemaphoreAwaitOp>(
+      loc, builder.getIntegerType(32), waitSemaphore, waitValue);
+  // TODO(scotttodd): if waitOp's result is not 0 (ok), fail
```

Comment:
Good question - vm ops *should* be ok here but I haven't tested that. If we wanted to avoid that we could create another dialect/op that was later turned into the VM ops (like `hal.check_success %status` -> cmp+cond_fail). I'm fine with either for now (so long as we have something checking).

Reply:
Iirc, there may be an overly zealous verifier in play for some VM ops (verifies that they are in a VM module). That verifier (if it still exists) probably just needs to go. We should be able to mix low level VM ops in if needed, imo.

Reply:
```
  func @main$async(%arg0: !hal.semaphore, %arg1: index, %arg2: !hal.buffer_view, %arg3: !hal.buffer_view, %arg4: !hal.semaphore, %arg5: index) -> !hal.buffer_view {
    %0 = hal.semaphore.await %arg0, min_value = %arg1 : i32
    %nz = vm.cmp.nz.i32 %0 : i32
    %c13 = vm.const.i32 13 : i32  // 13 == IREE_STATUS_INTERNAL
    vm.cond_fail %nz, %c13
    %1 = call @main$sync(%arg2, %arg3) : (!hal.buffer_view, !hal.buffer_view) -> !hal.buffer_view
    hal.semaphore.signal %arg4, value = %arg5
    return %1 : !hal.buffer_view
  }
```

fails with `error: 'vm.cond_fail' op expects parent op 'vm.func'` . I can take a look at adding `hal.check_success` instead, or this is an "overly zealous verifier"?

Reply:
Added `hal.check_success` with a lowering to `vm.cond_fail`.